Add programs.ssh module to generate .ssh/{config,authorized_keys}

league opened the pull request:
This doesn't embed ALL the options for the ssh client, but I based it
on options I use regularly. Here's an example usage extracted and
edited from my home.nix:

```nix
  programs.ssh = {
    enable = true;
    forwardAgent = true;
    controlMaster = "auto";
    authorizedKeys = [
      ssh/lydrik.pub
      ssh/lucian.pub
    ];
    hosts = {
      "something.blah.edu" = {
        port = 1024;
        user = "cleague";
        identitiesOnly = true;
      };
      "host1 host2 host2.net host2.com" = {
        port = 7422;
        hostname = "example.com";
        serverAliveInterval = 60;
      };
      "lucian" = {
        forwardX11 = true;
        forwardX11Trusted = true;
        checkHostIP = false;
      };
    };
  };
```

league commented:
Caveat: one problem I just ran into is that the generated `authorized_keys` doesn't actually work because it needs to be in a not-world-readable directory…

    debug1: Remote: Ignored authorized keys: bad ownership or modes for directory /nix/store

Maybe I could just remove the `authorizedKeys` option for now.

rycee commented:
This is really nice! At some point I started working on an [openssh module](https://gist.github.com/rycee/b11cbd6ddb140996450031bf2cfe0b50) where I started at the key end. I think your approach is more sane :-) Although, I still think having the keys in my Home Manager config would be cool.

In any case I like your module very much. I'll make some initial comments inline but unfortunately I don't have time to complete today.

rycee commented on modules/programs/ssh.nix:
```diff
@@ -0,0 +1,155 @@
+
+  hostStr = host: cf: concatStringsSep "\n" (
+    ["host ${host}"]
+    ++ optionals (cf.port != null)         ["  port ${toString cf.port}"]
```
Instead of "port", "forwardX11", and so on in the generated configuration I would prefer "Port", "ForwardX11", etc. since that is the capitalization used in the OpenSSH manual.

rycee replied:
Instead of, e.g.,
```
optionals (cf.port != null)         ["  port ${toString cf.port}"]
```
could use
```
optional (cf.port != null)         "  port ${toString cf.port}"
```
Not a huge difference, exactly, but maybe a little simpler :-)

league replied:
Oh yeah, that makes sense… I was thinking `optional` would generate a string instead of a singleton list, but there's `optionalString` for that.

rycee commented on modules/programs/ssh.nix:
```diff
@@ -0,0 +1,155 @@
+    };
+
+    hosts = mkOption {
+      type = types.loaOf hostModule;
```
I think this shoud be `types.attrsOf` rather than `types.loaOf` since the attribute name is important, i.e., I think it wouldn't work very well if you actually put a list of attributes here.

(I realize I've done the same mistake in the `gnome-terminal` module, will make a commit to fix that)

league replied:
Haha, I didn't know how to specify this type so I grepped around for usages, and maybe found it in gnome-terminal — though it wasn't even clear to me what "loa" means! 

Will address these, and cut the authorized keys part, but probably tomorrow.

league commented:
I updated this to the recommendations with a `push -f`. I also removed the `authorized_keys` stuff. It would be nice to specify the keys with nix, as you noted, but due to the permissions issue we'd need a way to create the file that's not just a symlink to the nix store. (Or we need to wait for the "private files in nix store" solution.)

rycee commented:
Was trying this out and sadly realized a fatal flaw in using an attribute set for the host clauses. An attribute set does not have a guaranteed order and unfortunately the SSH file relies heavily on ordering.

For example
```
Host foo.example.org
    User myoverriddenuser

Host *.example.org
    User myregularuser
```
is very different from
```
Host *.example.org
    User myregularuser

Host foo.example.org
    User myoverriddenuser
```

My suggestion is to change `programs.ssh.hosts` to be a `types.listOf hostModule` and add
```nix
      host = mkOption {
        type = types.str;
        example = "*.example.org";
        description = "The host pattern used by this clause.";
      };
```
to the `hostModule` definition.

Your example then becomes
```nix
  programs.ssh = {
    enable = true;
    forwardAgent = true;
    controlMaster = "auto";
    hosts = [
      {
        host = "something.blah.edu";
        port = 1024;
        user = "cleague";
        identitiesOnly = true;
      }

      {
        host = "host1 host2 host2.net host2.com";
        port = 7422;
        hostname = "example.com";
        serverAliveInterval = 60;
      }

      {
        host = "lucian";
        forwardX11 = true;
        forwardX11Trusted = true;
        checkHostIP = false;
      }
    ];
  };
```

Not quite as elegant syntax, unfortunately. On the other hand it is somewhat nice in that it may allow a future expansion to handle [Match](http://man.openbsd.org/OpenBSD-current/man5/ssh_config.5#Match) rules, e.g.,
```nix
{
    match = "host example.org exec \"test %p = 42\"";
    user = "foo";
}
```
but then the name `programs.ssh.hosts` may not be the best, perhaps `programs.ssh.rules`?

league commented:
Ah, good catch. I haven't been using wildcards at all, definitely not match. I could attempt a revision in a few days (schedule right now is tight) or feel free to take a crack yourself.

rycee commented:
I implemented the change I outlined above and pushed to your branch (since you had marked it pushable). Please have a look and see if you agree.

league commented:
LGTM. I pulled the change and revised my config to match; everything works. Feel free to squash (or not) and merge.

rycee commented:
Sqashed and rebased into master bce262e46e573c031feb968358ff31db77b00e71. I also cleaned up the commit message and option descriptions a little. Many thanks for your contribution!

ardumont commented:
> Caveat: one problem I just ran into is that the generated authorized_keys doesn't actually work because it needs to be in a not-world-readable directory…

Maybe a tad violent but setting `StrictModes no` in the sshd server configuration allows the authorized_keys file to be installed in the nix store (even though world readable), client side.

man sshd:
```
If this file, the ~/.ssh directory, or the user's home directory are writable by other users, then the file could be modified or replaced by unauthorized users.  In this case, sshd will not allow it to be used unless the StrictModes option has been set to “no”.
```

That may not be as violent as i thought.

Still, mabye a similar ssh client option exists, i did not check yet.
